### PR TITLE
Define non-panicking UTF encoding methods on `char`

### LINKED ISSUE #52579: Define non-panicking UTF encoding methods on `char`
i.e. `pub fn encode_utf8(self, &mut [u8]) -> Option<&mut str>` which returns `None` if buffer too short

PR to follow

### PR DESCRIPTION
**Pull request by strake**
Resolves #52579.

### CODE REVIEW

**Comment by oli-obk** [comment]
Since one needs to check the return value for `None` anyway, what's the motivation for this change?

You can write

```rust
if buf.len() >= c.utf8_len() {
    let s = c.encode_utf8(buf);
    // work with str
} else {
    // alternate action
}
```

instead of

```rust
if let Some(s) = c.try_encode_utf8(buf) {
    // work with str
} else {
    // alternate action, but cannot access `buf` here until nll are a thing
}
```

Do you have a real world usage example that can be used as motivation for this change?

Your PR message should also reference the corresponding issue that this PR is closing

**Comment by sfackler** [comment]
I'm also interested in the motivation for these methods.

**Comment by strake** [comment]
In general, i think the `try_encode` API (as defined in this PR) is more difficult to mess up and panic with; i needn't remember to check any lengths. It seems analogous to [`[_]::index`](https://doc.rust-lang.org/stable/std/ops/trait.Index.html#tymethod.index) vs [`[_]::get`](https://doc.rust-lang.org/stable/std/primitive.slice.html#method.get).

My use case is passing an encoding function to another function which fills a buffer and writes it out when full:
```rust
extern crate io;

use core::mem;
pub use self::io::*;

fn fst2<S, T>((x, _): (S, T)) -> S { x }

pub fn writeCode<Codon: Copy, F: FnMut(T, &mut [Codon]) -> Option<usize>, I: Iterator<Item = T>, T, W: Write<Codon>>
                (mut encode: F, w: &mut W, xs : I) -> Result<usize, W::Err> where W::Err: From<EndOfFile> {
    let mut buf: [Codon; 4096] = unsafe { mem::uninitialized() };
    let mut pos = 0;
    let mut nBytesWritten = 0;
    for x in xs {
        match encode(x, &mut buf[pos..]) {
            Some(n) => { pos += n; },
            None => {
                try!(w.write_all(&buf[0..pos]).map_err(fst2));
                nBytesWritten += pos;
                pos = 0;
            }
        }
    }
    try!(w.write_all(&buf[0..pos]).map_err(fst2));
    nBytesWritten += pos;
    Ok(nBytesWritten)
}
```

It is much more convenient to take a function `encode` which returns `Option` (or `Result`) rather than separate functions to encode and compute length.

File: src/libcore/char/methods.rs [hunk 1]
```diff
@@ -525,6 +556,37 @@ impl char {
...
+    ///
+    /// ```
+    /// let mut b = [0; 1];
+    ///
+    /// assert_eq!(None, '𝕊'.encode_utf16(&mut b));
+    /// ```
+    #[unstable(feature = "try_unicode_encode_char", issue = "52579")]
+    #[inline]
+    pub fn try_encode_utf16(self, dst: &mut [u16]) -> Option<&mut [u16]> {
+        if dst.len() < self.len_utf16() { None } else { Some(self.encode_utf16(dst)) }
```
Comment (oli-obk) [inline src/libcore/char/methods.rs]: so this is essentially doing the checks twice in the happy path, maybe move the unsafe code into this function and have the panicky function match on the result and panic on None?

File: src/libcore/char/methods.rs [hunk 2]
```diff
@@ -467,6 +467,37 @@ impl char {
...
+    ///
+    /// ```
+    /// let mut b = [0; 1];
+    ///
+    /// assert_eq!(None, 'ß'.encode_utf8(&mut b));
+    /// ```
+    #[unstable(feature = "try_unicode_encode_char", issue = "52579")]
+    #[inline]
+    pub fn try_encode_utf8(self, dst: &mut [u8]) -> Option<&mut str> {
+        if dst.len() < self.len_utf8() { None } else { Some(self.encode_utf8(dst)) }
```
Comment (oli-obk) [inline src/libcore/char/methods.rs]: same for this function

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/libcore/char/methods.rs (hunk 1 above)
Comment (strake) [reply src/libcore/char/methods.rs]: Done

**Comment by strake** [comment]
Bikeshed: ought this to return `Result<_, UtfError>` rather than `Option` (for some hypothetical type `UtfError`)? It might simplify the panicking methods, but i'm not sure whether it's more idiomatic.

**Comment by TimNN** [comment]
Ping from triage, @sfackler. This PR requires your review.

Also, @strake it looks like your PR needs to update some doc tests.

**Comment by SimonSapin** [comment]
I agree with @oli-obk that the motivation doesn’t feel enough to add new methods. It seems easy enough to check the buffer length against `c.utf8_len()`. And if you need to interoperate with some existing `Option`- or `Result`-based API, it seems easy enough to write a three-lines wrapper that does the `utf8_len` check.

You can even omit the length check entirely and use a 4-bytes buffer:

```rust
something.write_str(c.encode_utf8(&mut [0_u8; 4]))
```

So unless new arguments come up I’m inclined to not accept this:

@rfcbot fcp close

**Comment by strake** [comment]
@SimonSapin Unfortunately [that](http://play.rust-lang.org/?gist=30770eab7e09032b0f6e8bfcb113e482&version=stable&mode=release&edition=2015) emits all the panicking code, which i'd prefer to avoid.

**Comment by strake** [comment]
It's not a huge deal — i can rather define an extension trait (for example) — but extension traits are cumbersome, and it seems unfortunate to reinvent this functionality.

**Pull request closed without merging**